### PR TITLE
Accurate pressure formulation

### PR DESCRIPTION
**Pull request by arvigj**

### CODE REVIEW

File: src/polyfem/solver/AdjointNLProblem.cpp [hunk 1]
```diff
@@ -175,6 +297,41 @@ namespace polyfem::solver
 
 	bool AdjointNLProblem::is_step_valid(const Eigen::VectorXd &x0, const Eigen::VectorXd &x1)
 	{
+		Eigen::MatrixXd X, V0, V1, V_smooth;
+		Eigen::MatrixXi F;
+
+		int state_count = 0;
+		for (auto state_ : all_states_)
```
Comment (Huangzizhou) [inline src/polyfem/solver/AdjointNLProblem.cpp:304]: This should skip states that are not associated with any shape param

File: src/polyfem/assembler/MooneyRivlin3ParamElasticity.hpp [hunk 2]
```diff
@@ -36,13 +36,13 @@ namespace polyfem::assembler
 
 			const T J = polyfem::utils::determinant(def_grad);
 			const T log_J = log(J);
+			const auto F_tilde = def_grad;
```
Comment (Huangzizhou) [inline src/polyfem/assembler/MooneyRivlin3ParamElasticity.hpp:39]: Need to wait the other PR for these changes

File: src/polyfem/mesh/SlimSmooth.cpp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+#include <igl/boundary_facets.h>
+#include <polyfem/utils/Logger.hpp>
+#include <igl/slim.h>
+
+namespace polyfem::mesh
+{
+	namespace
+	{
+		double triangle_jacobian(const Eigen::VectorXd &v1, const Eigen::VectorXd &v2, const Eigen::VectorXd &v3)
```
Comment (Huangzizhou) [inline src/polyfem/mesh/SlimSmooth.cpp:11]: These functions also appear in AdjointNLProblem, can be moved to a header

File: src/polyfem/solver/forms/adjoint_forms/AdjointForm.cpp [hunk 4]
```diff
@@ -97,6 +97,12 @@ namespace polyfem::solver
 		return term;
 	}
 
+	void StaticForm::solution_changed(const Eigen::VectorXd &new_x)
+	{
+		AdjointForm::solution_changed(new_x);
+		solution_changed_step(0, new_x);
```
Comment (Huangzizhou) [inline src/polyfem/solver/forms/adjoint_forms/AdjointForm.cpp:103]: "solution_changed_step" doesn't seem to be necessary, can we change it back to just "solution_changed" and don't call it for every time step?

File: src/polyfem/solver/forms/adjoint_forms/BarrierForms.cpp [hunk 5]
```diff
@@ -87,7 +103,9 @@ namespace polyfem::solver
 			collision_mesh_,
 			collision_mesh_.vertices(V0),
 			collision_mesh_.vertices(V1),
-			broad_phase_method_, 1e-6, 1e6);
+			broad_phase_method_, dmin_, 1e-6, 1e6);
+
+		adjoint_logger().info("Objective {}: max step size is {}.", name(), max_step);
```
Comment (Huangzizhou) [inline src/polyfem/solver/forms/adjoint_forms/BarrierForms.cpp:108]: debug level

File: src/polyfem/solver/forms/adjoint_forms/TransientForm.cpp [hunk 6]
```diff
@@ -180,7 +180,7 @@ namespace polyfem::solver
 	void TransientForm::solution_changed(const Eigen::VectorXd &new_x)
 	{
 		AdjointForm::solution_changed(new_x);
-		obj_->solution_changed(new_x);
+		// obj_->solution_changed(new_x);
```
Comment (Huangzizhou) [inline src/polyfem/solver/forms/adjoint_forms/TransientForm.cpp:183]: Like mentioned above, can we remove "solution_changed_step" and use "solution_changed" instead?

File: src/polyfem/solver/AdjointNLProblem.cpp [hunk 7]
```diff
@@ -18,6 +19,125 @@ namespace polyfem::solver
...
+			{
+				for (const auto &state : p->get_states())
+					if (state.get() != curr_state.get())
+						continue;
+				if (p->get_parameter_type() != ParameterType::Shape)
+					continue;
+				auto state_variable = p->get_parametrization().eval(x);
+				auto output_indexing = p->get_output_indexing(x);
+				for (int i = 0; i < output_indexing.size(); ++i)
+					X(output_indexing(i)) = state_variable(i);
```
Comment (Huangzizhou) [inline src/polyfem/solver/AdjointNLProblem.cpp:135]: Just FYI, this could be simplified to "X(output_indexing) = state_variable;"

File: src/polyfem/solver/AdjointNLProblem.cpp [hunk 8]
```diff
@@ -261,6 +418,14 @@ namespace polyfem::solver
 	{
 		bool need_rebuild_basis = false;
 
+		std::vector<Eigen::MatrixXd> V_old_list;
```
Comment (Huangzizhou) [inline src/polyfem/solver/AdjointNLProblem.cpp:421]: should skip if no shape param

File: src/polyfem/solver/AdjointNLProblem.cpp [hunk 9]
```diff
@@ -282,6 +470,14 @@ namespace polyfem::solver
...
```
Comment (Huangzizhou) [inline src/polyfem/solver/AdjointNLProblem.cpp:473]: same here

File: src/polyfem/solver/AdjointNLProblem.cpp [hunk 10]
```diff
@@ -290,6 +486,26 @@ namespace polyfem::solver
 				need_rebuild_basis = true;
 		}
 
+		// Apply slim to all states on a frequency
+		if (need_rebuild_basis && enable_slim && curr_x.size() > 0)
```
Comment (Huangzizhou) [inline src/polyfem/solver/AdjointNLProblem.cpp:490]: Why not remove the slim from "is_step_valid" and only keep it here? If solution_changed throws an error, the line search would reduce step size and retry.

File: src/polyfem/solver/Optimizations.cpp [hunk 11]
```diff
@@ -87,7 +89,21 @@ namespace polyfem::solver
 			for (const auto &arg : args)
 				forms.push_back(create_form(arg, var2sim, states));
 
+			// bool amips_form = false;
```
Comment (Huangzizhou) [inline src/polyfem/solver/Optimizations.cpp:92]: remove comments?

File: src/polyfem/solver/Optimizations.cpp [hunk 12]
```diff
@@ -133,10 +149,77 @@ namespace polyfem::solver
...
+				}
+				tmp->set_reference(target_displacement);
+				obj = tmp;
+			}
 			else if (type == "center-target")
 			{
 				obj = std::make_shared<BarycenterTargetForm>(var2sim, args, states[args["state"]], states[args["target_state"]]);
 			}
+			else if (type == "sdf-target")
+			{
```
Comment (Huangzizhou) [inline src/polyfem/solver/Optimizations.cpp:178]: these are long, can they be moved to constructors?

File: src/polyfem/state/StateDiff.cpp [hunk 13]
```diff
@@ -295,9 +295,9 @@ namespace polyfem
 
 				if (sol_step == force_step - 1)
 				{
-					StiffnessMatrix body_force_hessian(u.size(), u.size());
-					solve_data.body_form->hessian_wrt_u_prev(u_prev, force_step * dt, body_force_hessian);
-					hessian_prev += body_force_hessian;
+					// StiffnessMatrix body_force_hessian(u.size(), u.size());
```
Comment (Huangzizhou) [inline src/polyfem/state/StateDiff.cpp:298]: Why?

File: CMakeLists.txt [hunk 14]
```diff
@@ -26,49 +28,52 @@ endif()
...
 if(POLYFEM_WITH_CCACHE AND CCACHE_PROGRAM)
-  message(STATUS "Enabling Ccache support")
-  set(ccacheEnv
-    CCACHE_BASEDIR=${CMAKE_BINARY_DIR}
-    CCACHE_SLOPPINESS=clang_index_store,include_file_ctime,include_file_mtime,locale,pch_defines,time_macros
-  )
-  foreach(lang IN ITEMS C CXX)
-    set(CMAKE_${lang}_COMPILER_LAUNCHER
-      ${CMAKE_COMMAND} -E env ${ccacheEnv} ${CCACHE_PROGRAM}
+    message(STATUS "Enabling Ccache support")
```
Comment (teseoch) [inline CMakeLists.txt:39]: are these necessary?

File: CMakeLists.txt [hunk 15]
```diff
@@ -26,49 +28,52 @@ endif()
...
+    set(ccacheEnv
+        CCACHE_BASEDIR=${CMAKE_BINARY_DIR}
+        CCACHE_SLOPPINESS=clang_index_store,include_file_ctime,include_file_mtime,locale,pch_defines,time_macros
     )
-  endforeach()
+
+    foreach(lang IN ITEMS C CXX)
```
Comment (teseoch) [inline CMakeLists.txt:45]: here too?

File: CMakeLists.txt [hunk 16]
```diff
@@ -224,6 +230,10 @@ target_link_libraries(polyfem PUBLIC simple_bvh::simple_bvh)
 include(mshio)
 target_link_libraries(polyfem PUBLIC mshio)
 
+# Nanospline
```
Comment (teseoch) [inline CMakeLists.txt:233]: do we need it?

File: cmake/recipes/nanospline.cmake [hunk 17]
```diff
@@ -0,0 +1,13 @@
+# Nanospline (https://github.com/qnzhou/nanospline)
+# License: Mozilla Public License 2.0
+
+if(TARGET nanospline)
```
Comment (teseoch) [inline cmake/recipes/nanospline.cmake:4]: necesary?

File: json-specs/input-spec.json [hunk 18]
```diff
@@ -1225,7 +1225,7 @@
     },
     {
         "pointer": "/contact/use_convergent_formulation",
-        "default": false,
+        "default": true,
```
Comment (teseoch) [inline json-specs/input-spec.json:1228]: scary

File: json-specs/input-spec.json [hunk 19]
```diff
@@ -2106,6 +2126,7 @@
             "value"
         ],
         "optional": [
+            "time_reference",
```
Comment (teseoch) [inline json-specs/input-spec.json:2129]: can we separate this?

File: json-specs/input-spec.json [hunk 20]
```diff
@@ -2151,6 +2183,44 @@
         ],
         "doc": "Neumann boundary condition"
     },
+    {
+        "pointer": "/boundary_conditions/rest_pressure_boundary",
```
Comment (teseoch) [inline json-specs/input-spec.json:2187]: i dont understand

File: json-specs/input-spec.json [hunk 21]
```diff
@@ -2151,6 +2183,44 @@
...
+        "default": [],
+        "type": "list",
+        "doc": "Neumann boundary condition for normal times value for vector-valued PDEs."
+    },
+    {
+        "pointer": "/boundary_conditions/rest_pressure_boundary/*",
```
Comment (teseoch) [inline json-specs/input-spec.json:2193]: is this the new pressure? do you like the name?

File: json-specs/objective-spec.json [hunk 22]
```diff
@@ -98,6 +107,44 @@
         ],
         "doc": "TODO"
     },
+    {
+        "pointer": "/",
```
Comment (teseoch) [inline json-specs/objective-spec.json:111]: do we need these?

File: src/polyfem/assembler/GenericProblem.cpp [hunk 23]
```diff
@@ -828,6 +860,8 @@ namespace polyfem
 						for (size_t k = 0; k < ff.size(); ++k)
 						{
 							displacements_[i].value[k].init(ff[k]);
+							if (j_boundary[i - offset].contains("time_reference") && j_boundary[i - offset]["time_reference"].size() > 0)
```
Comment (teseoch) [inline src/polyfem/assembler/GenericProblem.cpp:863]: can we separate this?

File: src/polyfem/assembler/PressureAssembler.cpp [hunk 24]
```diff
@@ -0,0 +1,831 @@
...
+
+							normals.row(n) = normals.row(n) * trafo.inverse();
+							normals.row(n).normalize();
+
+							if (mesh_.is_volume())
+							{
+								Eigen::Vector3d g1, g2, g3;
+								auto endpoints = utils::BoundarySampler::tet_local_node_coordinates_from_face(lb[i]);
+								g1 = trafo * (endpoints.row(0) - endpoints.row(1)).transpose();
+								g2 = trafo * (endpoints.row(0) - endpoints.row(2)).transpose();
```
Comment (teseoch) [inline src/polyfem/assembler/PressureAssembler.cpp:141]: what is this code?

File: src/polyfem/assembler/PressureAssembler.cpp [hunk 25]
```diff
@@ -0,0 +1,831 @@
...
+								}
+								g_3_grad_local.row(0) *= -1;
+
+								for (int k = 0; k < size_; ++k)
+									for (int l = 0; l < vals.basis_values.size() * size_; ++l)
+										g_3_grad(n, k * (vals.basis_values.size() * size_) + l) = g_3_grad_local(k, l);
+							}
+						}
+
+						if (multiply_pressure)
```
Comment (teseoch) [inline src/polyfem/assembler/PressureAssembler.cpp:597]: what is this?

File: src/polyfem/assembler/PressureAssembler.hpp [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+		public:
+			ThermodynamicProcess() {}
+			virtual ~ThermodynamicProcess(){};
+			virtual double pressure(const double start_pressure, const double v0, const double v1) const { return 0; }
+			virtual double first_derivative(const double start_pressure, const double v0, const double v1) const { return 0; }
+			virtual double second_derivative(const double start_pressure, const double v0, const double v1) const { return 0; }
+			virtual double energy(const double start_pressure, const double v0, const double v1) const { return 0; }
+
+		protected:
+			const double atmospheric = 101.3e3;
```
Comment (teseoch) [inline src/polyfem/assembler/PressureAssembler.hpp:25]: where are they used/changed? what about the unit?

File: src/polyfem/assembler/Problem.cpp [hunk 27]
```diff
@@ -22,22 +22,29 @@ namespace polyfem
 			std::vector<LocalBoundary> &local_boundary,
 			std::vector<int> &boundary_nodes,
 			std::vector<LocalBoundary> &local_neumann_boundary,
+			std::vector<LocalBoundary> &local_pressure_boundary,
```
Comment (teseoch) [inline src/polyfem/assembler/Problem.cpp:25]: this is becoming a mess :) we should thing about refactor it

File: src/polyfem/assembler/RhsAssembler.cpp [hunk 28]
```diff
@@ -626,18 +626,17 @@ namespace polyfem
...
-					if (!has_samples)
-						continue;
-
 					global_primitive_ids.setConstant(weights.size(), primitive_global_id);
+
 					vals.compute(e, mesh_.is_volume(), points, bs, gbs);
 
 					for (int n = 0; n < vals.jac_it.size(); ++n)
 					{
+						Eigen::MatrixXd ppp(1, size_);
```
Comment (teseoch) [inline src/polyfem/assembler/RhsAssembler.cpp:637]: why these changes?

File: src/polyfem/assembler/RhsAssembler.cpp [hunk 29]
```diff
@@ -626,18 +626,17 @@ namespace polyfem
 				for (int i = 0; i < lb.size(); ++i)
 				{
 					const int primitive_global_id = lb.global_primitive_id(i);
-					const auto nodes = bs.local_nodes_for_primitive(primitive_global_id, mesh_);
```
Comment (teseoch) [inline src/polyfem/assembler/RhsAssembler.cpp:629]: why changing?

File: src/polyfem/io/OutData.cpp [hunk 30]
```diff
@@ -92,9 +92,36 @@ namespace polyfem::io
...
+				// for (int n = 0; n < normals.rows(); ++n)
+				// {
+				// 	normals.row(n) = normals.row(n) * vals.jac_it[n];
+				// 	normals.row(n).normalize();
+				// }
+
 				for (int n = 0; n < normals.rows(); ++n)
 				{
-					normals.row(n) = normals.row(n) * vals.jac_it[n];
+					trafo = vals.jac_it[n].inverse();
```
Comment (teseoch) [inline src/polyfem/io/OutData.cpp:103]: what this not already fixed?

File: src/polyfem/mesh/SlimSmooth.cpp [hunk 31]
```diff
@@ -0,0 +1,134 @@
+#include "SlimSmooth.hpp"
+
+#include <igl/boundary_facets.h>
+#include <polyfem/utils/Logger.hpp>
+#include <igl/slim.h>
```
Comment (teseoch) [inline src/polyfem/mesh/SlimSmooth.cpp:5]: is this part of this pr?

File: src/polyfem/solver/forms/PressureForm.cpp [hunk 32]
```diff
@@ -0,0 +1,94 @@
...
+
+#include <polyfem/solver/AdjointTools.hpp>
+
+#include <polyfem/utils/Logger.hpp>
+
+namespace polyfem::solver
+{
+	namespace
+	{
+		class LocalThreadVecStorage
```
Comment (teseoch) [inline src/polyfem/solver/forms/PressureForm.cpp:21]: where is this used?

**Comment by teseoch** [comment]
i feel that this should be more than one pr, would it be possible to split it?

File: src/polyfem/solver/AdjointNLProblem.cpp [hunk 33]
```diff
@@ -18,6 +19,125 @@ namespace polyfem::solver
 {
 	namespace
 	{
+		double triangle_jacobian(const Eigen::VectorXd &v1, const Eigen::VectorXd &v2, const Eigen::VectorXd &v3)
```
[the remark this answers is not in the document]
Comment (arvigj) [reply src/polyfem/solver/AdjointNLProblem.cpp]: done

File: src/polyfem/solver/AdjointNLProblem.cpp (hunk 1 above)
Comment (arvigj) [reply src/polyfem/solver/AdjointNLProblem.cpp:215]: done

File: src/polyfem/assembler/MooneyRivlin3ParamElasticity.hpp (hunk 2 above)
Comment (arvigj) [reply src/polyfem/assembler/MooneyRivlin3ParamElasticity.hpp:39]: agreed

File: src/polyfem/mesh/SlimSmooth.cpp (hunk 3 above)
Comment (arvigj) [reply src/polyfem/mesh/SlimSmooth.cpp]: done

File: src/polyfem/solver/AdjointNLProblem.cpp (hunk 9 above)
Comment (arvigj) [reply src/polyfem/solver/AdjointNLProblem.cpp:384]: It only does caching in general, otherwise this is unused. I can refactor to remove but it would make the function longer.

File: src/polyfem/state/StateDiff.cpp (hunk 13 above)
Comment (arvigj) [reply src/polyfem/state/StateDiff.cpp]: fixed, had removed this type of energy before

File: src/polyfem/assembler/RhsAssembler.cpp (hunk 28 above)
Comment (arvigj) [reply src/polyfem/assembler/RhsAssembler.cpp]: removed

File: src/polyfem/assembler/RhsAssembler.cpp (hunk 29 above)
Comment (arvigj) [reply src/polyfem/assembler/RhsAssembler.cpp:629]: It's best to keep the declaration before the use of the variables, improves the readability.

File: src/polyfem/io/OutData.cpp (hunk 30 above)
Comment (arvigj) [reply src/polyfem/io/OutData.cpp:97]: It is fixed for the output but not for computing the continuous traction forces. In any case, the continuous traction forces are too inaccurate to use.
